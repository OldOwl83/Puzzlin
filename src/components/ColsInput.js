import { useContext, useState, useEffect } from "react";
import { PuzzleGrid } from "../RompeCocosApp";

export const ColsInput = () => {

    const { puzzleGrid, setPuzzleGrid } = useContext( PuzzleGrid );
    
    const [ input, setInput ] = useState( puzzleGrid[0] );
    
    useEffect(() => {
        
        if( input >= 3 || input == '' )
        {
            if( input >= 3 )
                setPuzzleGrid( [ Number( input ), puzzleGrid[1] ] );
            else
                setPuzzleGrid( [ Number( 3 ), puzzleGrid[1] ] );
        }else
        {
            setPuzzleGrid( [ Number( 3 ), puzzleGrid[1] ] );
            setInput( 3 );
        }
    }, [ input, setPuzzleGrid ]);

    return <input 
                type="number" 
                value={ input }
                min="3"
                onChange={ ( e ) => setInput( e.target.value) } 
            />
}
