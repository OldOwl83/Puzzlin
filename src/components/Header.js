import React from 'react';
import { Link, useLinkClickHandler } from 'react-router-dom';

export const Header = () => {

    const linker = useLinkClickHandler( '/', { 
        target: '_self',
        replace: false,
        state: {
            contScroll: 0,
            selectedPhoto: 'images/portada1.png',
            author: 'Mauro Donnantuoni',
        }
    });

    return (
        <>
        <Link to='/' id='toHomeLink' onClick={ linker }>
            <header>
                <h1>Puzzlin</h1>
                <div id="logoContainer">
                    <img src="images/rompecabezas.png" alt="logo" />
                </div>
            </header>
        </Link>

        <hr />
        </>
    );
};
