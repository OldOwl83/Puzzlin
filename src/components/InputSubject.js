import { useState } from "react";
import { PropTypes } from 'prop-types';

import '../styles/InputSubject.css';

export const InputSubject = ( { setSearch } ) => {

    console.log( 'InputSubject' );

    const [ input, setInput ] = useState( '' );
    
    const submitInput = (e) => {
        e.preventDefault();
        
        setInput( '' );

        setSearch( input );
    }

    return (

        <h2>Input a subject and select a photo. Then, press on the preview to start.
            <form onSubmit={ submitInput }>
                <input 
                    id="inputSubject"
                    type="text"
                    value={ input }
                    onChange={ (e) => setInput( e.target.value ) }
                />
            </form>
        </h2>
    )
}

InputSubject.propTypes = {

    setSearch: PropTypes.func.isRequired,
}