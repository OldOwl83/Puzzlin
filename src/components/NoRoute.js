import React from 'react';

export const NoRoute = () => {
  return <p className='message'>Nonexistent path.</p>;
};
