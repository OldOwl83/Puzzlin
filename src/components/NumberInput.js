/*
* General purpose number input. It receives by parameters a function which changes an external value or state, the minimum and maximum numbers allowed and the initial value. If the user tries to exceed the set range, it automatically sets the minimum or maximum accordingly. It sets the value as an integer data type.
*/

import { useState, useEffect, memo } from "react";

export const NumberInput = memo(( { setValue, min, max, initialValue }) => {

    const [ input, setInput ] = useState( initialValue );
    
    useEffect(() => {
    
        if( ( input >= min && input <= max ) || input === '' ) //It must include '' to allow erase manually the text in the input box.
        {
            if( input >= min && input <= max )
                setValue( Number( input ) );
            else 
                setValue( Number( min ) ); // If the input box remain empty, the set value is the minimum.
        }else if( input > max )
        {
            setValue( Number( max ) );
            setInput( max );
        }else
        {
            setValue( Number( min ) )
            setInput( min );
        }
    }, [ input, min, max, setValue ]);

    return (
    
        <input 
            type="number" 
            value={ input }
            min={ min }
            max={ max }
            onChange={ ( e ) => setInput( e.target.value ) } 
        />
    );
});
