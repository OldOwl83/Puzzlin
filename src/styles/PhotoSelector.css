h2{
    color: rgb(18, 112, 116);
    font-size: 1.2rem;
}

#selectorContainer{
    display: flex;
    justify-content:space-between;

    align-items: center;

    width: 96%;
    margin: 6px auto;
    padding: 14px;
}

#photosContainer{

    display: flex;
    flex-direction: column;
    justify-content: space-around;
    align-items: center;
    
    width: 17%;
    height: 540px;
    padding: 10px;
    
    border: 2px solid gray;
    border-radius: 10px;
    box-shadow: 2px 2px 4px black;
    background-color: rgb(245, 186, 186);
}

#photosContainer img{
    width: 88%;
    margin: 6px auto;
    
    border: 2px solid gray;
    border-radius: 10px;
    box-shadow: 3px 3px 4px black;
    
    transition: box-shadow 0.4s ease;
}

#photosContainer img:hover{
    box-shadow: none;
}

#imgPreview{
    width: 70%;
}

.message{
    display: inline-block;

    width: 70%;

    font-size: 34px;

    color: rgb(45, 22, 88);
}

