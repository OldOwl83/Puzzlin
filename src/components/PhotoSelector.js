import { useState, useEffect } from "react";

import { InputSubject } from "./InputSubject";
import { Preview } from './Preview';
import { useFetchPhotos } from "../hooks/useFetchPhotos";
import { PhotoShowcase } from "./PhotoShowcase";

import '../styles/PhotoSelector.css';
import 'animate.css';

export const PhotoSelector = ( { setMainPhoto } ) => {

    console.log('PhotoSelector');

    const [search, setSearch] = useState( '' );

    const [ imgSelected, setImgSelected] = useState( "images/tapa.png" );

    const { photos, loading } = useFetchPhotos( search );

    useEffect(() => {
        if(photos.length > 0)
            setImgSelected( photos[0].urlRegular );
    }, [ photos ])

    return (
        <>
            <InputSubject 
                setSearch={ setSearch } 
             />

            <div id="selectorContainer">
                
                <PhotoShowcase photos={ photos } setImage={ setImgSelected } />

                { loading && <p className="message">Loading...</p> }
                    
                { !photos && <p className="message">Connection to images provider failed.</p> }

                { search !== '' && !loading && photos.length < 1 && <p className="message">There were no matches with the search words.</p> }

                { (search === '' || photos.length > 0) && !loading && <Preview photo={ imgSelected } start={ setMainPhoto } /> }
            </div>
        </>
    )
}