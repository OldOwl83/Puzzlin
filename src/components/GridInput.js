/*
* It handles the contextual values of the puzzle grid. These values represent the columns and rows to cut out the puzzle photo and are in an array with two indices, which is a state belonging to the <Puzzlin /> component.
*/

import { useContext, useState, useEffect } from "react";

import { PuzzleGrid } from "../Puzzlin";
import { NumberInput } from "./NumberInput";

import "../styles/GridInput.css";

export const GridInput = () => {

    const { puzzleGrid, setPuzzleGrid } = useContext( PuzzleGrid );
    
    const [ cols, setCols ] = useState( puzzleGrid[0] );
    const [ rows, setRows ] = useState( puzzleGrid[1] );
    
    useEffect(() => {
        
        setPuzzleGrid( [ cols, rows ] );
        
    }, [ setPuzzleGrid, cols, rows ]);

    return (
    
        <div id="GridInput">
            <p>Columns
                <NumberInput setValue={ setCols } min={ 4 } max={ 8 } initialValue={ puzzleGrid[0] } />
                {/* It send de previous context values to initial value of this component, to maintain them across all routes. */}
            </p>
            <p>Rows
                <NumberInput setValue={ setRows } min={ 3 } max={ 8 } initialValue={ puzzleGrid[1] } />
            </p>
        </div>
    );
}
