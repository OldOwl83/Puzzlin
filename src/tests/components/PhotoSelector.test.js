import { shallow } from 'enzyme';
import { renderHook } from '@testing-library/react-hooks';

import { PhotoSelector } from "../../components/PhotoSelector";
import { useState } from 'react';


describe('Test on PhotoSelector component.', () => {

    const setMP = jest.fn();

    
    test('It should match to snapshot.', () => {
        
        const wrapper = shallow( <PhotoSelector setMainPhoto={ setMP } />);

        expect( wrapper ).toMatchSnapshot();
    })

    test('It should deployed the <img /> elements', () => {

        const imgs = [
            {
                id: 'anything',
                description: 'anything', 
                urlRegular: 'anything',
                urlSmall: 'anything',
                author: 'anything',
            },
            {
                id: 'anything',
                description: 'anything', 
                urlRegular: 'anything',
                urlSmall: 'anything',
                author: 'anything',
            },
            {
                id: 'anything',
                description: 'anything', 
                urlRegular: 'anything',
                urlSmall: 'anything',
                author: 'anything',
            },
            {
                id: 'anything',
                description: 'anything', 
                urlRegular: 'anything',
                urlSmall: 'anything',
                author: 'anything',
            },
            {
                id: 'anything',
                description: 'anything', 
                urlRegular: 'anything',
                urlSmall: 'anything',
                author: 'anything',
            },
            {
                id: 'anything',
                description: 'anything', 
                urlRegular: 'anything',
                urlSmall: 'anything',
                author: 'anything',
            },
            {
                id: 'anything',
                description: 'anything', 
                urlRegular: 'anything',
                urlSmall: 'anything',
                author: 'anything',
            },
            {
                id: 'anything',
                description: 'anything', 
                urlRegular: 'anything',
                urlSmall: 'anything',
                author: 'anything',
            },
            {
                id: 'anything',
                description: 'anything', 
                urlRegular: 'anything',
                urlSmall: 'anything',
                author: 'anything',
            },
            {
                id: 'anything',
                description: 'anything', 
                urlRegular: 'anything',
                urlSmall: 'anything',
                author: 'anything',
            },
        ];

        const args = {
            photos: imgs, 
            loading: false,
            connected: true,
        };

        const respu = renderHook( () => useState( { photos: imgs, loading: false, connected: true } ));

        // console.log( respu.result.current );

        const wrapper = shallow( <PhotoSelector setMainPhoto={ setMP } />);

        expect( wrapper ).toMatchSnapshot();
        // expect( wrapper.find( 'img' ).length ).toBeGreaterThan( 2 );
        
    })
    

    
})
