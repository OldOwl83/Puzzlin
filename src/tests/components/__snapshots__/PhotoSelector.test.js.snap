// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on PhotoSelector component. It should deployed the <img /> elements 1`] = `
<div
  id="photoSelector"
>
  <h2>
    Input a subject and select a photo. Then, press on the preview to start.
  </h2>
  <InputSubject
    setConnectedParent={[Function]}
    setLoadingParent={[Function]}
    setPhotosParent={[Function]}
  />
  <div
    id="photosContainer"
  >
    <p
      className="message"
    >
      There were no matches with the search words.
    </p>
  </div>
</div>
`;

exports[`Test on PhotoSelector component. It should match to snapshot. 1`] = `
<div
  id="photoSelector"
>
  <h2>
    Input a subject and select a photo. Then, press on the preview to start.
  </h2>
  <InputSubject
    setConnectedParent={[Function]}
    setLoadingParent={[Function]}
    setPhotosParent={[Function]}
  />
  <div
    id="photosContainer"
  >
    <p
      className="message"
    >
      There were no matches with the search words.
    </p>
  </div>
</div>
`;
